src/StreamJsonRpc.Tests/JsonRpcWithFatalExceptionsTests.cs

```diff
@@ -0,0 +1,304 @@
+        Assert.True(((DisposingMessageHandler)this.clientRpc.MessageHandler).IsDisposed);
+    }
+
+    public class Server
+    {
+        internal const string ThrowAfterCancellationMessage = "Throw after cancellation";
+
+        public bool DelayAsyncMethodWithCancellation { get; set; }
+
+        public AsyncAutoResetEvent ServerMethodReached { get; } = new AsyncAutoResetEvent();
+
+        public AsyncAutoResetEvent AllowServerMethodToReturn { get; } = new AsyncAutoResetEvent();
+
+        public static string ServerMethod(string argument)
+        {
+            return argument + "!";
+        }
+
+        public Task ServerMethodThatReturnsCancelledTask()
+        {
+            var tcs = new TaskCompletionSource<object>();
+            tcs.SetCanceled();
+            return tcs.Task;
+        }
+
+        public void MethodThatThrowsUnauthorizedAccessException(string message)
+        {
+            throw new UnauthorizedAccessException(message);
+        }
+
+        public async Task AsyncMethodThatThrowsAfterYield(string message)
+        {
+            await Task.Yield();
+            throw new Exception(message);
+        }
+
+        public Task AsyncMethodThatThrowsBeforeYield(string message)
+        {
+            var tcs = new TaskCompletionSource<object>();
+            tcs.SetException(new Exception(message));
+            return tcs.Task;
+        }
+
+        public async Task<string> AsyncMethodThatReturnsStringAndThrows(string message)
+        {
+            await Task.Run(() => throw new Exception(message));
+
+            return "never will return";
+        }
+
+        public async Task<string> AsyncMethodWithCancellation(string arg, CancellationToken cancellationToken)
+        {
+            this.ServerMethodReached.Set();
+
+            // TODO: remove when https://github.com/Microsoft/vs-threading/issues/185 is fixed
+            if (this.DelayAsyncMethodWithCancellation)
+            {
+                await Task.Delay(UnexpectedTimeout).WithCancellation(cancellationToken);
+            }
+
+            await this.AllowServerMethodToReturn.WaitAsync(cancellationToken);
+            return arg + "!";
+        }
+
+        public async Task<string> AsyncMethodFaultsAfterCancellation(string arg, CancellationToken cancellationToken)
+        {
+            this.ServerMethodReached.Set();
+            await this.AllowServerMethodToReturn.WaitAsync();
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                var cancellationSignal = new AsyncManualResetEvent();
+                using (cancellationToken.Register(() => cancellationSignal.Set()))
+                {
+                    await cancellationSignal;
+                }
+            }
+
+            throw new InvalidOperationException(ThrowAfterCancellationMessage);
+        }
+
+        public async Task MethodThatThrowsAsync(string message)
+        {
+            await Task.Run(() => throw new Exception(message));
```

Comment:
I've already implemented that in AsyncMethodThatThrowsAfterYield. I'll remove this method and it's corresponding tests since it doesn't seem to be doing anything unique.